pstoremem/protobook.go

```diff
@@ -0,0 +1,143 @@
+package pstoremem
+
+import (
+	"sync"
+
+	peer "github.com/libp2p/go-libp2p-peer"
+
+	pstore "github.com/libp2p/go-libp2p-peerstore"
+)
+
+const (
+	maxInternedProtocols    = 512
+	maxInternedProtocolSize = 256
+)
+
+type protoSegment struct {
+	sync.RWMutex
+	protocols map[peer.ID]map[string]struct{}
+}
+
+type protoSegments [256]*protoSegment
+
+func (s *protoSegments) get(p peer.ID) *protoSegment {
+	b := []byte(p)
+	return s[b[len(b)-1]]
```

Comment:
Go may optimize this, but it might just decide to copy the slice. How about `s[byte(p[len(p)-1])]`?

Reply:
hrm, I don't think that it copies in such casts, but hard to be certain.

Reply:
How does the rune to byte conversion work here btw?

Reply:
> hrm, I don't think that it copies in such casts, but hard to be certain.

It `[]byte(string)` usually copies as `[]byte` is mutable whyle strings aren't. However, go does the right thing in _some_ cases, IIRC.

> How does the rune to byte conversion work here btw?

Go indexes strings by byte. Think of go strings like python strings, they can contain arbitrary bytes but are sometimes valid UTF8.

Reply:
ok, avoided the intermediate byte slice.